File: packages/core/src/records.ts
import type { RandomProvenance } from "./rng/types.js";

/**
 * Version of the serialized event shape. Additive, optional fields keep the
 * version; renaming, removing, or re-meaning fields bumps it and requires a
 * migration note (ADR-0006).
 *
 * Version 2 (ADR-0015) widened `sides` to any face count and `value` to any
 * integer, so a custom die can record faces that are not 1..N, and added the
 * optional `die` and `label`. Version 1 records still deserialize: their
 * values are already valid version 2, and reading one returns a version 2
 * record with the same numbers.
 */
export const EVENT_SCHEMA_VERSION = 2;

/** Schema versions this core can read. */
export const SUPPORTED_SCHEMA_VERSIONS: readonly number[] = [1, 2];

/** One physical die inside a roll, in rolled order. */
export type DieOutcome = {
  /** How many faces the die has. */
  readonly sides: number;
  /**
   * What the die contributes: the face rolled (1..sides) for a plain numeric
   * die, or the face's value for a custom one, which may be negative or zero.
   */
  readonly value: number;
  /** False when a keep/drop selection removed this die from the subtotal. */
  readonly kept: boolean;
  /** Custom die this came from, when it was not a plain numeric die. */
  readonly die?: string;
  /** How the face reads, when that differs from the value. */
  readonly label?: string;
};

/** The outcome of one dice group term such as "2d20kh1". */
export type RollGroupOutcome = {
  /** Canonical notation for this group, e.g. "2d20kh1", "4d{fate}". */
  readonly notation: string;
  readonly sign: 1 | -1;
  /** Face count of every die in the group. */
  readonly sides: number;
  /** Custom die name, when the group rolled one. */
  readonly die?: string;
  readonly dice: readonly DieOutcome[];
  /** Sum of kept die values, before the sign is applied. */
  readonly subtotal: number;
};

export type RollResult = {
  readonly kind: "roll";
  readonly schemaVersion: number;
  /** Canonical form of the rolled expression, e.g. "2d20kh1+3". */
  readonly expression: string;
  readonly groups: readonly RollGroupOutcome[];
  /** Net signed contribution of all constant terms. */
  readonly modifier: number;
  readonly total: number;
  readonly provenance: RandomProvenance;
};

export type CoinFlipResult = {
  readonly kind: "coin-flip";
  readonly schemaVersion: number;
  readonly outcome: "heads" | "tails";
  readonly provenance: RandomProvenance;
};

/** Every event record the core can produce. */
export type InteractionEvent = RollResult | CoinFlipResult;

/** Recursively freezes a record so resolved outcomes are tamper-evident. */
export function deepFreeze<T>(value: T): T {
  if (value !== null && typeof value === "object") {
    for (const key of Object.keys(value)) {
      deepFreeze((value as Record<string, unknown>)[key]);
    }
    Object.freeze(value);
  }
  return value;
}

File: adapters/unity/Runtime/Records.cs
using System.Collections.Generic;
using System.Globalization;
using System.Text;

namespace DiceForge
{
    /// <summary>
    /// Minimal JSON writing, so a record serializes to exactly the schema v2
    /// shape every DiceForge platform reads — optional keys omitted rather than
    /// written as null, which is what makes the vectors comparable.
    /// </summary>
    public static class Json
    {
        /// <summary>A JSON string literal, escaped like JSON.stringify would.</summary>
        public static string Quote(string value)
        {
            var builder = new StringBuilder("\"");
            foreach (char c in value)
            {
                switch (c)
                {
                    case '"': builder.Append("\\\""); break;
                    case '\\': builder.Append("\\\\"); break;
                    case '\b': builder.Append("\\b"); break;
                    case '\f': builder.Append("\\f"); break;
                    case '\n': builder.Append("\\n"); break;
                    case '\r': builder.Append("\\r"); break;
                    case '\t': builder.Append("\\t"); break;
                    default:
                        if (c < 0x20) builder.Append("\\u").Append(((int)c).ToString("x4"));
                        else builder.Append(c);
                        break;
                }
            }
            return builder.Append('"').ToString();
        }

        public static string Number(int value) => value.ToString(CultureInfo.InvariantCulture);
    }

    /// <summary>One physical die inside a roll, in rolled order.</summary>
    public sealed class DieOutcome
    {
        public int Sides;
        public int Value;
        public bool Kept;

        /// <summary>Custom die this came from, when it was not a plain numeric die.</summary>
        public string Die;

        /// <summary>How the face reads, when that differs from the value.</summary>
        public string Label;

        /// <summary>"reroll" or "explosion" when the die is beyond the count asked for.</summary>
        public string Source;

        /// <summary>True when a reroll threw this result away (ADR-0016).</summary>
        public bool Rerolled;

        public string ToJson()
        {
            var builder = new StringBuilder("{");
            builder.Append("\"sides\":").Append(Json.Number(Sides));
            builder.Append(",\"value\":").Append(Json.Number(Value));
            builder.Append(",\"kept\":").Append(Kept ? "true" : "false");
            if (Die != null) builder.Append(",\"die\":").Append(Json.Quote(Die));
            if (Label != null) builder.Append(",\"label\":").Append(Json.Quote(Label));
            if (Source != null) builder.Append(",\"source\":").Append(Json.Quote(Source));
            if (Rerolled) builder.Append(",\"rerolled\":true");
            return builder.Append('}').ToString();
        }
    }

    /// <summary>The outcome of one dice group term such as "2d20kh1".</summary>
    public sealed class RollGroupOutcome
    {
        public string Notation;
        public int Sign;
        public int Sides;
        public string Die;
        public List<DieOutcome> Dice = new List<DieOutcome>();
        public int Subtotal;

        public string ToJson()
        {
            var builder = new StringBuilder("{");
            builder.Append("\"notation\":").Append(Json.Quote(Notation));
            builder.Append(",\"sign\":").Append(Json.Number(Sign));
            builder.Append(",\"sides\":").Append(Json.Number(Sides));
            if (Die != null) builder.Append(",\"die\":").Append(Json.Quote(Die));
            builder.Append(",\"dice\":[");
            for (int index = 0; index < Dice.Count; index++)
            {
                if (index > 0) builder.Append(',');
                builder.Append(Dice[index].ToJson());
            }
            builder.Append(']');
            builder.Append(",\"subtotal\":").Append(Json.Number(Subtotal));
            return builder.Append('}').ToString();
        }
    }

    /// <summary>What every resolved event shares.</summary>
    public abstract class InteractionEvent
    {
        /// <summary>Version of the serialized shape; 2 today (ADR-0015).</summary>
        public const int SchemaVersion = 2;

        public Provenance Provenance;

        public abstract string ToJson();

        protected string ProvenanceJson()
        {
            var builder = new StringBuilder("{");
            builder.Append("\"source\":").Append(Json.Quote(Provenance.Source));
            if (Provenance.Seed != null) builder.Append(",\"seed\":").Append(Json.Quote(Provenance.Seed));
            builder.Append(",\"algorithm\":").Append(Json.Quote(Provenance.Algorithm));
            return builder.Append('}').ToString();
        }
    }

    public sealed class RollResult : InteractionEvent
    {
        /// <summary>Canonical form of the rolled expression, e.g. "2d20kh1+3".</summary>
        public string Expression;

        public List<RollGroupOutcome> Groups = new List<RollGroupOutcome>();

        /// <summary>Net signed contribution of all constant terms.</summary>
        public int Modifier;

        public int Total;

        public override string ToJson()
        {
            var builder = new StringBuilder("{");
            builder.Append("\"kind\":\"roll\"");
            builder.Append(",\"schemaVersion\":").Append(Json.Number(SchemaVersion));
            builder.Append(",\"expression\":").Append(Json.Quote(Expression));
            builder.Append(",\"groups\":[");
            for (int index = 0; index < Groups.Count; index++)
            {
                if (index > 0) builder.Append(',');
                builder.Append(Groups[index].ToJson());
            }
            builder.Append(']');
            builder.Append(",\"modifier\":").Append(Json.Number(Modifier));
            builder.Append(",\"total\":").Append(Json.Number(Total));
            builder.Append(",\"provenance\":").Append(ProvenanceJson());
            return builder.Append('}').ToString();
        }
    }

    public sealed class CoinFlipResult : InteractionEvent
    {
        /// <summary>"heads" or "tails".</summary>
        public string Outcome;

        public override string ToJson()
        {
            var builder = new StringBuilder("{");
            builder.Append("\"kind\":\"coin-flip\"");
            builder.Append(",\"schemaVersion\":").Append(Json.Number(SchemaVersion));
            builder.Append(",\"outcome\":").Append(Json.Quote(Outcome));
            builder.Append(",\"provenance\":").Append(ProvenanceJson());
            return builder.Append('}').ToString();
        }
    }
}
